providers: add template provider

josharian opened the pull request:
Fixes #215.

This PR is an initial pass, so that I can get early feedback. At a minimum, it lacks website/documentation updates, but I want to do that last, after the code has stabilized and the barns have been painted.

Sample `example.tf`:

```
variable "key" {
    default = "hello"
}

variable "filename" {
    default = "template.txt"
}

resource "template_file" "sample" {
    filename = "${var.filename}"
    vars {
        hello = "${var.key}"
        world = "${replace(var.key, "ello", "i")}"
    }
}

output "rendered" {
    value = "${template_file.sample.rendered}"
}
```

Sample `template.txt`:

```
Hello: ${hello}
World: ${world}
Addition: ${1+2+3}
```

Output:

```
$ terraform apply
var.filename
  Default: template.txt
  Enter a value: 

var.key
  Default: hello
  Enter a value: 

template_file.sample: Creating...
  filename:   "" => "template.txt"
  rendered:   "" => "<computed>"
  vars.#:     "0" => "2"
  vars.hello: "" => "hello"
  vars.world: "" => "hi"
template_file.sample: Creation complete

Apply complete! Resources: 1 added, 0 changed, 0 destroyed.

The state of your infrastructure has been saved to the path
below. This state is required to modify and destroy your
infrastructure, so keep it safe. To inspect the complete state
use the `terraform show` command.

State path: terraform.tfstate

Outputs:

  rendered = Hello: hello
World: hi
Addition: 6
```

And everything shows up nicely for historical purposes in [`terraform.tfstate`](https://gist.github.com/josharian/1a1db45eadf5ba926f2f).

phinze commented on builtin/providers/template/resource.go:
```diff
@@ -0,0 +1,119 @@
+	// Reload every time in case something has changed.
+	// This should be cheap, and cache invalidation is hard.
+	return false, nil
+}
```
You can just omit `Exists`, which has the same effect.

phinze replied:
Oh nevermind - didn't see that `Read` was a noop. This makes sense.

phinze replied:
Hey @josharian - working on #1866 and coming back around to this. I think we need to rework the lifecycle a bit here. Will follow up with a PR.

josharian replied:
Ok. Sorry about that. Let me know if you'd like me to dive in. (Sounds from your comment like you have it under control.)

phinze commented on builtin/providers/template/resource.go:
```diff
@@ -0,0 +1,119 @@
+			"rendered": &schema.Schema{
+				Type:        schema.TypeString,
+				Computed:    true,
+				Default:     nil,
```
No need to call out `Default` here - it's already `nil`.

josharian replied:
Ack.

phinze commented on builtin/providers/template/resource.go:
```diff
@@ -0,0 +1,119 @@
+	filename := d.Get("filename").(string)
+	vars := d.Get("vars").(map[string]interface{})
+
+	buf, err := ioutil.ReadFile(filename)
```
Can you do a homedir expand on this?

See https://github.com/hashicorp/terraform/blob/master/config/interpolate_funcs.go#L62

josharian replied:
Will do. I'll address all commentary in a follow-up commit. If you like, I'll squash everything at the end, so we have a clean history.

phinze replied:
Sounds good! I'm a big git-spelunker so I'm pro-squash. But generally we're pretty flexible.

josharian commented:
Thanks! I'll leave this here for a few days so that others who care about this (@apparentlymart?) can weigh in, can assess whether this will address their needs, and can object to the names I used (template_file, filename, vars, etc.).

Meanwhile, since I didn't hear any showstopping concerns from you, I'll write some docs.

phinze commented:
This is a great start! :grinning: 

Inline'd a couple of nits. Think you could take a stab at some acceptance tests?

josharian commented:
> Think you could take a stab at some acceptance tests?

Absolutely.

apparentlymart commented:
This looks great to me. Thanks for taking this on, @josharian! This would be very useful to us if it got merged. (Our first use-case will be to templatize the user_data for our EC2 instances from a separate file, and it looks like this implementation will work just fine for that.)

josharian commented:
I've addressed the review comments, added tests, and written some docs. There's a CI failure, but it looks unrelated (consul timeout, vet failure in existing code). Ready for round two.

errordeveloper commented:
This is using Terraform's own string interpolation macros, isn't it? To me the idea of templates would really make most sense if templates would expose a reacher language, one with loops and logic, as means of going just a bit beyond what HCL currently offers.

errordeveloper commented:
It's great to be able to do the kind of file interpolation, which is what subject of #215 state and this covers it very well. As for #642, I wished for a more extended kind of language at the time of posting. I do suppose that #642 can be achieved with an external plugin, if the core codebase isn't going to provide much beyond interpolation macros in files.

phinze commented:
This looks great @josharian - merging so this will make it into 0.5 release this week. :tada: 

@errordeveloper - yep I think we'll definitely be talking about support for richer templating languages down the road, but this represents a great first step

josharian commented:
Just noticed that this didn't make it into the changelog. Want to do the honors, @phinze?

Thanks for all the help, and don't worry, the barrage of PRs from me should slow soon. :)

errordeveloper commented:
@phinze cool, I just created #1845 :+1:

errordeveloper commented:
@josharian thanks a million for getting this of the ground!

phinze commented:
@josharian ^^ done :+1:

c4milo commented:
This is great work. It would be extremely useful too if we can render resource's attributes into provisioners scripts without creating cyclic dependencies. I have my provisioners full of sed substitutions and would love to be able to render my scripts using templates instead. I believe this may require treating templates as first class citizens instead of resources. Any thoughts?

table-deploy commented:
Wow. This is fantastic. Thanks for writing this, @josharian!

nathanielks commented:
Wow. This is fantastic. Thanks for writing this, @josharian!
